### PR TITLE
Add support for macOS

### PR DESCRIPTION
**Pull request by muellermartin**
Compiling on macOS requires a few changes to the Makefile and some
really trivial changes to some source files (only white space!) because
Clang seems to be more strict to the standard.

First issue was resolving compile errors regarding missing newline
characters at the end of some source files as this is required by the
standard (error: no newline at end of file).

A slightly more complicated issue was Clang complaining about that
"ISO C requires a translation unit to contain at least one declaration"
which was caused by an empty source file because an #ifdef was false.
This issue was resolved by only adding these source files when they are
needed. Propably the defines (HAVE_MMX, HAVE_SSE, etc.) can be left out.

Lastly a check was introduced to set compiler flags and file extensions
required or specific for macOS.

Additionally a few smaller issues in the Makefile where resolved:
-  the "bin" folder is not created (perhaps this was not noticed as the
"clean" target does not remove the "bin" folder)
- the "clean" taget now removes the "bin" folder to get a cleaner state

I did not check if building still works on Linux. I suspect at least one change (line 137/158: change from ${LD} to ${CC}) could cause a little trouble. It would be nice if someone could check this out.

### CODE REVIEW

**Comment by zbeekman** [comment]
This still fails for me on Yosemite, any thoughts on how to fix, @muellermartin ?

```
mkdir -p obj_lib && cc -O2 -g -Wall -Werror -Wextra -pedantic -std=c99 -Ilib -DHAVE_MMX -DHAVE_SSE -DHAVE_AVX -DNO_HAVE_NEON -march=native -o obj_lib/sha1_simd_avx256.lo -c lib/sha1_simd_avx256.c
In file included from lib/sha1_simd_avx256.c:22:
lib/sha1_simd.cinc:38:10: error: implicit declaration of function '_mm256_or_si256' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                W[i] = SIMD_ROL_V(SIMD_XOR_VV( SIMD_XOR_VV(W[i - 3],W[i - 8]),SIMD_XOR_VV(W[i - 14],W[i - 16]) ), 1);
                       ^
lib/simd_avx256.h:85:26: note: expanded from macro 'SIMD_ROL_V'
#define SIMD_ROL_V(l,i)         SIMD_OR_VV(SIMD_SHL_V(l,i),SIMD_SHR_V(l,32-i))
                                ^
lib/simd_avx256.h:33:26: note: expanded from macro 'SIMD_OR_VV'
#define SIMD_OR_VV(l,r)         _mm256_or_si256(l,r)
                                ^
In file included from lib/sha1_simd_avx256.c:22:
lib/sha1_simd.cinc:38:10: error: implicit declaration of function '_mm256_slli_epi32' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
lib/simd_avx256.h:85:37: note: expanded from macro 'SIMD_ROL_V'
#define SIMD_ROL_V(l,i)         SIMD_OR_VV(SIMD_SHL_V(l,i),SIMD_SHR_V(l,32-i))
                                           ^
lib/simd_avx256.h:38:26: note: expanded from macro 'SIMD_SHL_V'
#define SIMD_SHL_V(l,i)         _mm256_slli_epi32(l,i)
                                ^
lib/simd_avx256.h:33:42: note: expanded from macro 'SIMD_OR_VV'
#define SIMD_OR_VV(l,r)         _mm256_or_si256(l,r)
                                                ^
In file included from lib/sha1_simd_avx256.c:22:
lib/sha1_simd.cinc:38:21: error: implicit declaration of function '_mm256_xor_si256' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                W[i] = SIMD_ROL_V(SIMD_XOR_VV( SIMD_XOR_VV(W[i - 3],W[i - 8]),SIMD_XOR_VV(W[i - 14],W[i - 16]) ), 1);
                                  ^
lib/simd_avx256.h:35:26: note: expanded from macro 'SIMD_XOR_VV'
#define SIMD_XOR_VV(l,r)        _mm256_xor_si256(l,r)
                                ^
lib/simd_avx256.h:85:48: note: expanded from macro 'SIMD_ROL_V'
#define SIMD_ROL_V(l,i)         SIMD_OR_VV(SIMD_SHL_V(l,i),SIMD_SHR_V(l,32-i))
                                                      ^
lib/simd_avx256.h:38:44: note: expanded from macro 'SIMD_SHL_V'
#define SIMD_SHL_V(l,i)         _mm256_slli_epi32(l,i)
                                                  ^
lib/simd_avx256.h:33:42: note: expanded from macro 'SIMD_OR_VV'
#define SIMD_OR_VV(l,r)         _mm256_or_si256(l,r)
                                                ^
lib/sha1_simd.cinc:38:21: note: did you mean '_mm256_or_si256'?
lib/simd_avx256.h:35:26: note: expanded from macro 'SIMD_XOR_VV'
#define SIMD_XOR_VV(l,r)        _mm256_xor_si256(l,r)
                                ^
lib/simd_avx256.h:85:48: note: expanded from macro 'SIMD_ROL_V'
#define SIMD_ROL_V(l,i)         SIMD_OR_VV(SIMD_SHL_V(l,i),SIMD_SHR_V(l,32-i))
                                                      ^
lib/simd_avx256.h:38:44: note: expanded from macro 'SIMD_SHL_V'
#define SIMD_SHL_V(l,i)         _mm256_slli_epi32(l,i)
                                                  ^
lib/simd_avx256.h:33:42: note: expanded from macro 'SIMD_OR_VV'
#define SIMD_OR_VV(l,r)         _mm256_or_si256(l,r)
                                                ^
lib/sha1_simd.cinc:38:10: note: '_mm256_or_si256' declared here
                W[i] = SIMD_ROL_V(SIMD_XOR_VV( SIMD_XOR_VV(W[i - 3],W[i - 8]),SIMD_XOR_VV(W[i - 14],W[i - 16]) ), 1);
                       ^
lib/simd_avx256.h:85:26: note: expanded from macro 'SIMD_ROL_V'
#define SIMD_ROL_V(l,i)         SIMD_OR_VV(SIMD_SHL_V(l,i),SIMD_SHR_V(l,32-i))
                                ^
lib/simd_avx256.h:33:26: note: expanded from macro 'SIMD_OR_VV'
#define SIMD_OR_VV(l,r)         _mm256_or_si256(l,r)
                                ^
In file included from lib/sha1_simd_avx256.c:22:
lib/sha1_simd.cinc:38:10: error: implicit declaration of function '_mm256_srli_epi32' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                W[i] = SIMD_ROL_V(SIMD_XOR_VV( SIMD_XOR_VV(W[i - 3],W[i - 8]),SIMD_XOR_VV(W[i - 14],W[i - 16]) ), 1);
                       ^
lib/simd_avx256.h:85:53: note: expanded from macro 'SIMD_ROL_V'
#define SIMD_ROL_V(l,i)         SIMD_OR_VV(SIMD_SHL_V(l,i),SIMD_SHR_V(l,32-i))
                                                           ^
lib/simd_avx256.h:39:26: note: expanded from macro 'SIMD_SHR_V'
#define SIMD_SHR_V(l,i)         _mm256_srli_epi32(l,i)
                                ^
lib/simd_avx256.h:33:44: note: expanded from macro 'SIMD_OR_VV'
#define SIMD_OR_VV(l,r)         _mm256_or_si256(l,r)
                                                  ^
lib/sha1_simd.cinc:38:10: note: did you mean '_mm256_slli_epi32'?
lib/simd_avx256.h:85:53: note: expanded from macro 'SIMD_ROL_V'
#define SIMD_ROL_V(l,i)         SIMD_OR_VV(SIMD_SHL_V(l,i),SIMD_SHR_V(l,32-i))
                                                           ^
lib/simd_avx256.h:39:26: note: expanded from macro 'SIMD_SHR_V'
#define SIMD_SHR_V(l,i)         _mm256_srli_epi32(l,i)
                                ^
lib/simd_avx256.h:33:44: note: expanded from macro 'SIMD_OR_VV'
#define SIMD_OR_VV(l,r)         _mm256_or_si256(l,r)
                                                  ^
lib/sha1_simd.cinc:38:10: note: '_mm256_slli_epi32' declared here
lib/simd_avx256.h:85:37: note: expanded from macro 'SIMD_ROL_V'
#define SIMD_ROL_V(l,i)         SIMD_OR_VV(SIMD_SHL_V(l,i),SIMD_SHR_V(l,32-i))
                                           ^
lib/simd_avx256.h:38:26: note: expanded from macro 'SIMD_SHL_V'
#define SIMD_SHL_V(l,i)         _mm256_slli_epi32(l,i)
                                ^
lib/simd_avx256.h:33:42: note: expanded from macro 'SIMD_OR_VV'
#define SIMD_OR_VV(l,r)         _mm256_or_si256(l,r)
                                                ^
In file included from lib/sha1_simd_avx256.c:22:
lib/sha1_simd.cinc:38:8: error: assigning to '__m256i' (vector of 4 'long long' values) from incompatible type 'int'
                W[i] = SIMD_ROL_V(SIMD_XOR_VV( SIMD_XOR_VV(W[i - 3],W[i - 8]),SIMD_XOR_VV(W[i - 14],W[i - 16]) ), 1);
                     ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lib/sha1_simd.cinc:47:8: error: assigning to '__m256i' (vector of 4 'long long' values) from incompatible type 'int'
                W[i] = SIMD_ROL_V(SIMD_XOR_VV( SIMD_XOR_VV(W[i - 3],W[i - 8]),SIMD_XOR_VV(W[i - 14],W[i - 16]) ), 1);
                     ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lib/sha1_simd.cinc:51:2: error: implicit declaration of function '_mm256_add_epi32' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        HASHCLASH_SHA1COMPRESS_ROUND1_STEP(a, b, c, d, e, W, 0);
        ^
lib/sha1_simd.cinc:18:8: note: expanded from macro 'HASHCLASH_SHA1COMPRESS_ROUND1_STEP'
        { e = SIMD_ADD_VV(e,SIMD_ADD_VV(sha1_f1(b,c,d),SIMD_ADD_VV(SIMD_ROL_V(a, 5),SIMD_ADD_VW(m[t],0x5A827999)))); b = SIMD_ROL_V(b, 30); }
              ^
lib/simd_avx256.h:27:26: note: expanded from macro 'SIMD_ADD_VV'
#define SIMD_ADD_VV(l,r)        _mm256_add_epi32(l,r)
                                ^
In file included from lib/sha1_simd_avx256.c:22:
lib/sha1_simd.cinc:51:2: error: implicit declaration of function '_mm256_and_si256' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
lib/sha1_simd.cinc:18:34: note: expanded from macro 'HASHCLASH_SHA1COMPRESS_ROUND1_STEP'
        { e = SIMD_ADD_VV(e,SIMD_ADD_VV(sha1_f1(b,c,d),SIMD_ADD_VV(SIMD_ROL_V(a, 5),SIMD_ADD_VW(m[t],0x5A827999)))); b = SIMD_ROL_V(b, 30); }
                                        ^
lib/sha1_simd.cinc:13:38: note: expanded from macro 'sha1_f1'
#define sha1_f1(b,c,d) SIMD_XOR_VV(d,SIMD_AND_VV(b,SIMD_XOR_VV(c,d)))
                                     ^
lib/simd_avx256.h:31:26: note: expanded from macro 'SIMD_AND_VV'
#define SIMD_AND_VV(l,r)        _mm256_and_si256(l,r)
                                ^
lib/simd_avx256.h:35:45: note: expanded from macro 'SIMD_XOR_VV'
#define SIMD_XOR_VV(l,r)        _mm256_xor_si256(l,r)
                                                   ^
lib/simd_avx256.h:27:43: note: expanded from macro 'SIMD_ADD_VV'
#define SIMD_ADD_VV(l,r)        _mm256_add_epi32(l,r)
                                                 ^
lib/simd_avx256.h:27:45: note: expanded from macro 'SIMD_ADD_VV'
#define SIMD_ADD_VV(l,r)        _mm256_add_epi32(l,r)
                                                   ^
In file included from lib/sha1_simd_avx256.c:22:
lib/sha1_simd.cinc:51:2: error: assigning to '__m256i' (vector of 4 'long long' values) from incompatible type 'int'
        HASHCLASH_SHA1COMPRESS_ROUND1_STEP(a, b, c, d, e, W, 0);
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lib/sha1_simd.cinc:18:6: note: expanded from macro 'HASHCLASH_SHA1COMPRESS_ROUND1_STEP'
        { e = SIMD_ADD_VV(e,SIMD_ADD_VV(sha1_f1(b,c,d),SIMD_ADD_VV(SIMD_ROL_V(a, 5),SIMD_ADD_VW(m[t],0x5A827999)))); b = SIMD_ROL_V(b, 30); }
            ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lib/sha1_simd.cinc:51:2: error: assigning to '__m256i' (vector of 4 'long long' values) from incompatible type 'int'
        HASHCLASH_SHA1COMPRESS_ROUND1_STEP(a, b, c, d, e, W, 0);
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lib/sha1_simd.cinc:18:113: note: expanded from macro 'HASHCLASH_SHA1COMPRESS_ROUND1_STEP'
        { e = SIMD_ADD_VV(e,SIMD_ADD_VV(sha1_f1(b,c,d),SIMD_ADD_VV(SIMD_ROL_V(a, 5),SIMD_ADD_VW(m[t],0x5A827999)))); b = SIMD_ROL_V(b, 30); }
                                                                                                                       ^ ~~~~~~~~~~~~~~~~~
lib/sha1_simd.cinc:52:2: error: assigning to '__m256i' (vector of 4 'long long' values) from incompatible type 'int'
        HASHCLASH_SHA1COMPRESS_ROUND1_STEP(e, a, b, c, d, W, 1);
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lib/sha1_simd.cinc:18:6: note: expanded from macro 'HASHCLASH_SHA1COMPRESS_ROUND1_STEP'
        { e = SIMD_ADD_VV(e,SIMD_ADD_VV(sha1_f1(b,c,d),SIMD_ADD_VV(SIMD_ROL_V(a, 5),SIMD_ADD_VW(m[t],0x5A827999)))); b = SIMD_ROL_V(b, 30); }
            ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lib/sha1_simd.cinc:52:2: error: assigning to '__m256i' (vector of 4 'long long' values) from incompatible type 'int'
        HASHCLASH_SHA1COMPRESS_ROUND1_STEP(e, a, b, c, d, W, 1);
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lib/sha1_simd.cinc:18:113: note: expanded from macro 'HASHCLASH_SHA1COMPRESS_ROUND1_STEP'
        { e = SIMD_ADD_VV(e,SIMD_ADD_VV(sha1_f1(b,c,d),SIMD_ADD_VV(SIMD_ROL_V(a, 5),SIMD_ADD_VW(m[t],0x5A827999)))); b = SIMD_ROL_V(b, 30); }
                                                                                                                       ^ ~~~~~~~~~~~~~~~~~
lib/sha1_simd.cinc:53:2: error: assigning to '__m256i' (vector of 4 'long long' values) from incompatible type 'int'
        HASHCLASH_SHA1COMPRESS_ROUND1_STEP(d, e, a, b, c, W, 2);
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lib/sha1_simd.cinc:18:6: note: expanded from macro 'HASHCLASH_SHA1COMPRESS_ROUND1_STEP'
        { e = SIMD_ADD_VV(e,SIMD_ADD_VV(sha1_f1(b,c,d),SIMD_ADD_VV(SIMD_ROL_V(a, 5),SIMD_ADD_VW(m[t],0x5A827999)))); b = SIMD_ROL_V(b, 30); }
            ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lib/sha1_simd.cinc:53:2: error: assigning to '__m256i' (vector of 4 'long long' values) from incompatible type 'int'
        HASHCLASH_SHA1COMPRESS_ROUND1_STEP(d, e, a, b, c, W, 2);
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lib/sha1_simd.cinc:18:113: note: expanded from macro 'HASHCLASH_SHA1COMPRESS_ROUND1_STEP'
        { e = SIMD_ADD_VV(e,SIMD_ADD_VV(sha1_f1(b,c,d),SIMD_ADD_VV(SIMD_ROL_V(a, 5),SIMD_ADD_VW(m[t],0x5A827999)))); b = SIMD_ROL_V(b, 30); }
                                                                                                                       ^ ~~~~~~~~~~~~~~~~~
lib/sha1_simd.cinc:54:2: error: assigning to '__m256i' (vector of 4 'long long' values) from incompatible type 'int'
        HASHCLASH_SHA1COMPRESS_ROUND1_STEP(c, d, e, a, b, W, 3);
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lib/sha1_simd.cinc:18:6: note: expanded from macro 'HASHCLASH_SHA1COMPRESS_ROUND1_STEP'
        { e = SIMD_ADD_VV(e,SIMD_ADD_VV(sha1_f1(b,c,d),SIMD_ADD_VV(SIMD_ROL_V(a, 5),SIMD_ADD_VW(m[t],0x5A827999)))); b = SIMD_ROL_V(b, 30); }
            ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lib/sha1_simd.cinc:54:2: error: assigning to '__m256i' (vector of 4 'long long' values) from incompatible type 'int'
        HASHCLASH_SHA1COMPRESS_ROUND1_STEP(c, d, e, a, b, W, 3);
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lib/sha1_simd.cinc:18:113: note: expanded from macro 'HASHCLASH_SHA1COMPRESS_ROUND1_STEP'
        { e = SIMD_ADD_VV(e,SIMD_ADD_VV(sha1_f1(b,c,d),SIMD_ADD_VV(SIMD_ROL_V(a, 5),SIMD_ADD_VW(m[t],0x5A827999)))); b = SIMD_ROL_V(b, 30); }
                                                                                                                       ^ ~~~~~~~~~~~~~~~~~
lib/sha1_simd.cinc:55:2: error: assigning to '__m256i' (vector of 4 'long long' values) from incompatible type 'int'
        HASHCLASH_SHA1COMPRESS_ROUND1_STEP(b, c, d, e, a, W, 4);
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lib/sha1_simd.cinc:18:6: note: expanded from macro 'HASHCLASH_SHA1COMPRESS_ROUND1_STEP'
        { e = SIMD_ADD_VV(e,SIMD_ADD_VV(sha1_f1(b,c,d),SIMD_ADD_VV(SIMD_ROL_V(a, 5),SIMD_ADD_VW(m[t],0x5A827999)))); b = SIMD_ROL_V(b, 30); }
            ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lib/sha1_simd.cinc:55:2: error: assigning to '__m256i' (vector of 4 'long long' values) from incompatible type 'int'
        HASHCLASH_SHA1COMPRESS_ROUND1_STEP(b, c, d, e, a, W, 4);
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lib/sha1_simd.cinc:18:113: note: expanded from macro 'HASHCLASH_SHA1COMPRESS_ROUND1_STEP'
        { e = SIMD_ADD_VV(e,SIMD_ADD_VV(sha1_f1(b,c,d),SIMD_ADD_VV(SIMD_ROL_V(a, 5),SIMD_ADD_VW(m[t],0x5A827999)))); b = SIMD_ROL_V(b, 30); }
                                                                                                                       ^ ~~~~~~~~~~~~~~~~~
lib/sha1_simd.cinc:56:2: error: assigning to '__m256i' (vector of 4 'long long' values) from incompatible type 'int'
        HASHCLASH_SHA1COMPRESS_ROUND1_STEP(a, b, c, d, e, W, 5);
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lib/sha1_simd.cinc:18:6: note: expanded from macro 'HASHCLASH_SHA1COMPRESS_ROUND1_STEP'
        { e = SIMD_ADD_VV(e,SIMD_ADD_VV(sha1_f1(b,c,d),SIMD_ADD_VV(SIMD_ROL_V(a, 5),SIMD_ADD_VW(m[t],0x5A827999)))); b = SIMD_ROL_V(b, 30); }
            ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
make: *** [obj_lib/sha1_simd_avx256.lo] Error 1
```

**Comment by muellermartin** [comment]
@zbeekman Thanks for the report. I have no instant idea why this fails, but I'll try to find out and help where I can.

**Comment by muellermartin** [comment]
@zbeekman Hm, I have to admit that I'm a total C noob. I stared at the error messages and the code for a bit and I conclude that it must either have something to do with your compiler being stricter (because of a proably earlier toolchain version) or (in my opinion more likely) that something with AVX2 goes wrong. The question is what exactly that could be. Does you CPU and OS support AVX2 (you can probably check this by running this command: `sysctl -a | grep machdep.cpu.leaf7_features` on Mac OS - [source](http://csharpmulticore.blogspot.de/2014/12/how-to-check-intel-avx2-support-on-mac-os-x-haswell.html))?

Another thing I stumbled upon was [this answer on almighty Stack Overflow](http://stackoverflow.com/a/20760453/1532986):
It basically suggests adding the compiler option `-Wa,-q`. You could test that by adding it to the variable `AVXFLAGS` in the `Makefile` (should be [line 88](https://github.com/cr-marcstevens/sha1collisiondetection/pull/4/commits/41a4da883d1861886ebf000eb06418d92e99224e#diff-b67911656ef5d18c4ae36cb6741b7965R88)).

**Comment by highcapital** [comment]
the -mavx2 option fixes the problem, thanks

**Comment by zbeekman** [comment]
> Does you CPU and OS support AVX2 (you can probably check this by running this command: sysctl -a | grep machdep.cpu.leaf7_features on Mac OS - source)?

Nope, only AVX 1 it appears

```
$ sysctl -a | grep -i avx
hw.optional.avx1_0: 1
hw.optional.avx2_0: 0
machdep.cpu.features: FPU VME DE PSE TSC MSR PAE MCE CX8 APIC SEP MTRR PGE MCA CMOV PAT PSE36 CLFSH DS ACPI MMX FXSR SSE SSE2 SS HTT TM PBE SSE3 PCLMULQDQ DTES64 MON DSCPL VMX SMX EST TM2 SSSE3 CX16 TPR PDCM SSE4.1 SSE4.2 x2APIC POPCNT AES PCID XSAVE OSXSAVE TSCTMR AVX1.0
```

This project needs a proper build system with introspection (using e.g., autotools or CMake) since there are certainly plenty of machines out there with AVX1 but not AVX2 to be truly portable.

**Comment by zbeekman** [comment]
> the -mavx2 option fixes the problem, thanks

@highcapital is correct, `-mavx2` is not getting added to the clang compile line for me.... investigating.

**Comment by zbeekman** [comment]
The issue is gmake vs make that ships with Mac OS/Xcode. After installing make with homebrew, and running gmake instead of make the problem goes away. Basically, there are sets of pattern rules with more specialized suffixes and gmake selects the most specific rule, but make that ships with latest Xcode on Yosemite does not, it defaults to the first, more generic rule it encounters, and therefore doesn't add any of the SIMD flags (SSE, AVX, etc.)

**Comment by muellermartin** [comment]
@zbeekman You're totally right. I did not notice that! I simply rearranged the more specific rule to be the last one and now it seems to work for me. Could you guys test it again?

**Comment by muellermartin** [comment]
I also noticed the efforts of @johnhawkinson to create a Homebrew Formula (a third party package manager for macOS) in his PR homebrew/homebrew-core#10431.

Basically these are the things we need to consider to get an acceptable package for the Homebrew maintainers:

- Add a tag for (pre-)release (see #11)
- Add an "install" target with the `PREFIX` option
- Probably add system feature detection to include/exclude the according source files based on actual CPU features (might be worth switching to autotools, cmake etc.)

**Comment by johnhawkinson** [comment]
I'm in some kind of Homebrew build environment hell right now, so I can't test trivial changes, but an install target should look something like this:

```
diff --git a/Makefile b/Makefile
index d4e1659..3ab1cd5 100644
--- a/Makefile
+++ b/Makefile
@@ -15,11 +15,14 @@ CC ?= gcc
 CFLAGS=-O2 -g -Wall -Werror -Wextra -pedantic -std=c99 -Ilib
 LDFLAGS=-O2 -g
 
-LT_CC:=libtool --tag=CC --mode=compile $(CC)
+LIBTOOL=libtool
+LT_CC:=$(LIBTOOL) --tag=CC --mode=compile $(CC)
 LT_CC_DEP:=$(CC)
-LT_LD:=libtool --tag=CC --mode=link $(CC)
+LT_LD:=$(LIBTOOL) --tag=CC --mode=link $(CC)
+LT_INSTALL:=$(LIBTOOL) --tag=CC --mode=install install
 
 MKDIR=mkdir -p
+PREFIX=/usr/local
 
 CC=${LT_CC}
 CC_DEP=${LT_CC_DEP}
@@ -178,3 +181,8 @@ ${LIB_OBJ_DIR}/%neon128.lo ${LIB_OBJ_DIR}/%neon128.o: ${LIB_DIR}/%neon128.c ${LI
 
 
 -include $(FS_DEP)
+
+install:
+       install -c bin/sha1dcsum $(PREFIX)/bin
+       install -c bin/sha1dcsum_partialcoll $(PREFIX)/bin
+       $(LT_INSTALL) -c bin/libdetectcoll.la $(PREFIX)/lib
```

(though for some reason this seems to try to find libdetectcoll.lai, but I have no reason to think my build environment is sane...)

**Comment by muellermartin** [comment]
@johnhawkinson Thanks for your efforts and the help with adding the install target! Based on your suggestion I added a new commit (c220376) adressing this issue. As an extra I also added the `uninstall` target.

Note: I've also changed `sha1dcsum_partialcoll` to be a symlink instead of a copy to save a few MB disk space ;)

**Comment by cr-marcstevens** [comment]
Hi all, thanks for the work on MacOS. Before I merge I have to check whether it doesn't break anything on linux and get someone local to confirm on MacOS. If you can make a PR for simpler commits on final endline, install and symlink? Those are easier to already confirm.

**Comment by muellermartin** [comment]
@cr-marcstevens Hi, thanks for the nice tool in the first place :)

I'll try to split up the changes, so it will be easier to test on Linux.

**Comment by muellermartin** [comment]
@cr-marcstevens I've split up parts of this PR to #13, #14 and #15. These changes should only affect the current build on Linux as the fixes for macOS are not included, yet.

Once these PRs are merged (or rejected) I'll try to create a new PR with just the changes related to macOS based on that state.

**Comment by cr-marcstevens** [comment]
Thanks a lot!
I'm currently setting up TravisCI, so we can more easily catch mistakes and also see how it compiles on MacOS.

File: Makefile [hunk 1]
```diff
@@ -12,12 +12,26 @@ endif
 TARGET ?= x86
 CC ?= gcc
 
+PREFIX=/usr/local
```
Comment (MikeMcQuaid) [inline Makefile]: Would be preferable to use `?=` here I think for packagers to be able to specify this.

File: Makefile [hunk 2]
```diff
@@ -12,12 +12,26 @@ endif
...
 LT_CC_DEP:=$(CC)
+UNAME:=$(shell uname)
+
+ifeq ($(UNAME),Darwin)
+LT_CC:=$(CC)
+LT_LD:=ld -dylib
+LDFLAGS=
+LIB_EXT=dylib
+else
+LT_CC:=libtool --tag=CC --mode=compile $(CC)
```
Comment (MikeMcQuaid) [inline Makefile]: A `LIBTOOL ?= libtool` would be nice for e.g. specifying `glibtool`

File: Makefile (hunk 1 above)
Comment (muellermartin) [inline Makefile]: I'll reflect this in PR #15.

Comment (MikeMcQuaid) [inline Makefile]: Great, thanks @muellermartin!

File: Makefile (hunk 2 above)
Comment (muellermartin) [inline Makefile]: As you can see I went another by using `ld -dylib` instead of `libtool` on macOS. Would you like a variable for LIBTOOL nevertheless?

Comment (MikeMcQuaid) [inline Makefile]: Ah, I see. If `libtool` can be avoided then fine as-is 👍

**Comment by muellermartin** [comment]
Like announced I created a new PR (#16) to add support for macOS (fewer changes because other issues were fixed in smaller PRs and commits).

I suggest this PR gets closed in favor of the new PR.

**Comment by cr-marcstevens** [comment]
Hi Martin,
Thanks for creating the new PR, I'm closing this.

**Pull request closed without merging**